server/service/microsoft_mdm.go

```diff
@@ -1188,31 +1194,104 @@ func (svc *Service) GetMDMWindowsTOSContent(ctx context.Context, redirectUri str
 }
 
 // isTrustedRequest checks if the incoming request was sent from MDM enrolled device
-func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML) error {
+func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML, reqCerts []*x509.Certificate) error {
 	if reqSyncML == nil {
 		return fleet.NewInvalidArgumentError("syncml req message", "message is not present")
 	}
 
-	// TODO - Security checks that need to be implemented
-	// - TLS based auth
-	// - Device auth based on Source/LocURI DeviceID information (this should be present on Enrollment DB)
+	// Checking if calling request is coming from an already MDM enrolled device
+	deviceID, err := reqSyncML.GetSource()
+	if err != nil || deviceID == "" {
+		return fmt.Errorf("invalid SyncML message %w", err)
+	}
+
+	enrolledDevice, err := svc.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
+	if err != nil || enrolledDevice == nil {
+		return errors.New("device was not MDM enrolled")
+	}
 
-	return nil
+	// Check if TLS certs contains device ID on its common name
+	if len(reqCerts) > 0 {
+		for _, reqCert := range reqCerts {
+			if strings.Contains(reqCert.Subject.CommonName, deviceID) {
+				return nil
+			}
+		}
+	}
+
+	// TODO: Latest version of the MDM client stack don't populate TLS.PeerCertificates array
+	// This is a temporary workaround to allow the management request to proceed
+	// Transport-level security should be replaced for Application-level security
+	// Transport-level security is defined in the MS-MDM spec in section 1.3.1
+	// On the other hand, Application-level security is defined here
+	// https://www.openmobilealliance.org/release/DM/V1_2_1-20080617-A/OMA-TS-DM_Security-V1_2_1-20080617-A.pdf
+	// The initial values for Application-level security configuration are defined in the
+	// WAP Profile blob that is sent to the device during the enrollment process. Example below
+	//	<characteristic type="APPAUTH">
+	//		<parm name="AAUTHLEVEL" value="CLIENT"/>
+	//		<parm name="AAUTHTYPE" value="DIGEST"/>
+	//		<parm name="AAUTHSECRET" value="2jsidqgffx"/>
+	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
+	//	</characteristic>
+	//	<characteristic type="APPAUTH">
+	//		<parm name="AAUTHLEVEL" value="APPSRV"/>
+	//		<parm name="AAUTHTYPE" value="DIGEST"/>
+	//		<parm name="AAUTHNAME" value="43f8bf591b8557346021"/>
+	//		<parm name="AAUTHSECRET" value="crbr3w2cab"/>
+	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
+	//	</characteristic>
+
+	if len(reqCerts) == 0 {
+		return nil
+	}
+
+	return errors.New("calling device is not trusted")
```

Comment:
this will be a `500` error by default. Can we make it an unauthorized request?

Reply:
This error is causing `GetMDMWindowsManagementResponse` to fail, and to return a SOAP fault to the MDM client stack, to indicate the calling OS that MDM that there was a problem with the ongoing management session  - see details [here](https://github.com/fleetdm/fleet/blob/b32c61b4a9b7b9b779d213d462ad41a0a7320273/server/service/microsoft_mdm.go#L883)

Reply:
thank you!